### ISSUE TITLE
Problems using this with custom elements

### ISSUE THREAD

**Issue by Thanood** [aurelia/ui-virtualization#45]
Hi,

I have some problems using this plugin with custom elements.
Using this markup:

``` html
<template>
  <mdl-list style="height: 120px; overflow: scroll;">
      <mdl-list-item lines.bind="showDescription ? 3 : 2" virtual-repeat.for="actor of actors">
        <mdl-list-item-primary>
          <span>${actor.name}</span>
          <span show.bind="!showDescription" class="mdl-list__item-sub-title">${actor.episodes} Episodes</span>
          <span show.bind="showDescription" class="mdl-list__item-text-body">
            ${actor.description}
          </span>
        </mdl-list-item-primary>
      </mdl-list-item>
    </mdl-list>
</template>
```

I'm getting this error:

```
utilities.js:17 Uncaught (in promise) TypeError: Cannot read property 'getBoundingClientRect' of null
    at calcOuterHeight (http://localhost:24562/jspm_packages/npm/aurelia-ui-virtualization@0.4.2/utilities.js:17:21)
    at VirtualRepeat._calcInitialHeights (http://localhost:24562/jspm_packages/npm/aurelia-ui-virtualization@0.4.2/virtual-repeat.js:389:56)
    at VirtualRepeat.itemsChanged (http://localhost:24562/jspm_packages/npm/aurelia-ui-virtualization@0.4.2/virtual-repeat.js:197:12)
    at VirtualRepeat.attached (http://localhost:24562/jspm_packages/npm/aurelia-ui-virtualization@0.4.2/virtual-repeat.js:139:12)
    at Controller.attached (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:2877:24)
    at View.attached (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1020:24)
    at View.attached (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1025:21)
    at ViewSlot.attached (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1403:15)
    at View.attached (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1025:21)
    at View.attached (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1025:21)
```

`firstViewElement` [here](https://github.com/aurelia/ui-virtualization/blob/master/src/virtual-repeat.js#L323) is null for me because `this.view(0).firstChild` in the line above is a comment (one of Aurelia's `<!-- anchor -->` comments).

In another view [this line](https://github.com/aurelia/ui-virtualization/blob/master/src/dom-helper.js#L8) fails for me because `element` seems to be the `#document-fragment`. The injected element (injected into virtual-repeat) is an `<!-- anchor -->` comment somehow (it's same setup as the list before, `virtual-repeat` is set on a custom element). Thus, the view-strategy returns `#document-fragment`

The error is different:

```
dom-helper.js:26 Uncaught (in promise) TypeError: Cannot read property 'overflowY' of undefined
    at DomHelper.hasOverflowScroll (http://localhost:24562/jspm_packages/npm/aurelia-ui-virtualization@0.4.2/dom-helper.js:26:19)
    at VirtualRepeat.attached (http://localhost:24562/jspm_packages/npm/aurelia-ui-virtualization@0.4.2/virtual-repeat.js:144:26)
    at Controller.attached (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:2877:24)
    at View.attached (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1020:24)
    at View.attached (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1025:21)
    at Controller.attached (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:2881:19)
    at View.attached (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1020:24)
    at ViewSlot.add (http://localhost:24562/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1272:14)
    at http://localhost:24562/jspm_packages/npm/aurelia-templating-router@1.0.0-beta.1.2.0/router-view.js:166:43
```

This is using a flexbox layout. I don't know if it's related but I'm writing it down for the issue to be as complete as possible. :smile:

**Comment by martingust** [comment]
Thanks for reporting this. I will take a look.

**Comment by martingust** [comment]
@Thanood Just tested your markup with master and is seems to have been resolved. Will go out in next release.

**Comment by Thanood** [comment]
Great! :+1: 
Thank you for looking into this.

**Comment by Thanood** [comment]
Unfortunately, this is not solved in my case. Maybe I'm doing something wrong.
Is there a gist.run I could possibly use to try or demonstrate?

I'll try to highlight the differences between using ui-virtualization with a custom element and a native element.

native:

``` html
    <div>
      <mdl-list-item virtual-repeat.for="item of items">
        <mdl-list-item-primary>
          <span>${item.title}</span>
        </mdl-list-item-primary>
      </mdl-list-item>
    </div>
```

generated html (only the first item):

``` html
<div>
    <div style="height: 0px;"/>
    <!--<view>-->
    <mdl-list-item class="au-target mdl-list__item" au-target-id="86">
        <!--<view>-->
        <mdl-list-item-primary class="au-target mdl-list__item-primary-content" au-target-id="85">
            <!--<view>-->
            <!--<view>-->
            <span>item 1</span>
            <!--</view>-->
            <!--anchor-->
            <!--</view>-->
        </mdl-list-item-primary>
        <!--anchor-->
        <!--anchor-->
        <!--</view>-->
    </mdl-list-item>
    <!--</view>-->

    ...

    <!--anchor-->
    <div style="height: 45840px;"/>
</div>
```

custom:

``` html
    <mdl-list>
      <mdl-list-item virtual-repeat.for="item of items">
        <mdl-list-item-primary>
          <span>${item.title}</span>
        </mdl-list-item-primary>
      </mdl-list-item>
    </mdl-list>
```

generated html:

``` html
<mdl-list class="au-target mdl-list" au-target-id="88">
    <!--<view>-->

    <mdl-list-item class="au-target mdl-list__item" au-target-id="86">
        <!--<view>-->
        <!-- <span class="mdl-list__item-primary-content">
    <content></content>
  </span>
  <content select=".mdl-list__item-secondary-content"></content> -->
        <mdl-list-item-primary class="au-target mdl-list__item-primary-content" au-target-id="85">
            <!--<view>-->
            <!--<view>-->
            <span>item 1</span>
            <!--</view>-->
            <!--anchor-->
            <!--</view>-->
        </mdl-list-item-primary>
        <!--anchor-->
        <!--anchor-->
        <!--</view>-->
    </mdl-list-item>
    <!--anchor-->
    <!--</view>-->
</mdl-list>
```

The injected viewSlot in the native version points to the containing `<div>` (the topmost in the example). See `anchor.parentNode`:

![image](https://cloud.githubusercontent.com/assets/677826/14789520/3ba1eb50-0b0d-11e6-8d5d-11676fe1e244.png)

(sorry, I failed to copy the console output)

The viewSlot of the custom element version points to the `#document-fragment` somehow. Again `anchor.parentNode`:

![image](https://cloud.githubusercontent.com/assets/677826/14789620/a7d4d102-0b0d-11e6-8fed-491b5c380b3a.png)

I cannot rule out that I'm doing something wrong but I don't see it.
Would it help if I modified a version of the skeleton navigation with some simple content and upload it here?

**Comment by Thanood** [comment]
Could it possibly be the use of `<content>` elements in mdl-list?
I'm not (explicitly) using ShadowDOM..

**Comment by Thanood** [comment]
@martingust I've been able to reproduce my problem here: https://github.com/Thanood/skeleton-ui-virtualization

The current version of [my-list.html](https://github.com/Thanood/skeleton-ui-virtualization/blob/master/src/features/my-list.html) reproduces the problem mentioned above:

``` html
<template>
  <content select="my-list-item"></content>
</template>
```

The first element is created, the viewSlot is injected as a child of `#document-fragment`. It leads to the exact same problem as explained above.

What's interesting is when you remove the `select` attribute from `<content>`.

``` html
<template>
  <content></content>
</template>
```

In this case `this.view(0).firstChild.parentNode` resolves to the `<my-list>` element and it works **for 70 list items** - in my case a little less than three pages.

The error issued in this case:

```
dom-helper.js:31 Uncaught (in promise) TypeError: Cannot read property 'overflowY' of undefined
    at DomHelper.hasOverflowScroll (http://localhost:9000/jspm_packages/github/aurelia/ui-virtualization@master/dom-helper.js:31:19)
    at VirtualRepeat.attached (http://localhost:9000/jspm_packages/github/aurelia/ui-virtualization@master/virtual-repeat.js:154:26)
    at Controller.attached (http://localhost:9000/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:2877:24)
    at View.attached (http://localhost:9000/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1020:24)
    at View.attached (http://localhost:9000/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1025:21)
    at ViewSlot.attached (http://localhost:9000/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1403:15)
    at View.attached (http://localhost:9000/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1025:21)
    at ViewSlot.attached (http://localhost:9000/jspm_packages/npm/aurelia-templating@1.0.0-beta.1.2.2/aurelia-templating.js:1403:15)
    at http://localhost:9000/jspm_packages/npm/aurelia-framework@1.0.0-beta.1.2.2/aurelia-framework.js:204:30
```

The `element` here:

``` javascript
    DomHelper.prototype.hasOverflowScroll = function hasOverflowScroll(element) {
      var style = element.style;
      return style.overflowY === 'scroll' || style.overflow === 'scroll' || style.overflowY === 'auto' || style.overflow === 'auto';
    };
```

is another `´#document-fragment``.

Removing `<my-list>` and repeating over `<my-list-item>`s kind of solves this (apart from the fact that in a "real" application you'd probably do something meaningful to `<my-list>` :smile:)

So.. I'm quite convinced that this is either an issue with nested content elements or with having `virtual-repeat` in _injected content_. That or I got `<content>` wrong.

What do you think?

edit: I tried this with master branch and npm release 0.4.3

**Comment by martingust** [comment]
@Thanood Thanks for repro and info. I will take a look at this and get back. If you want to dig in there and give it a shot to fix that would be greatly appreciated! I'll reopen this.

**Comment by Thanood** [comment]
@martingust Thank you for reopening and having a look again. :smile: 

I've started to try a fix but didn't get far yet. I only had this simple idea of creating another wrapper inside the shadow root (?) which acts as a new scrollContainer (kind of).

So that would become something like this.

``` html
<my-list>
  <!-- <content> is used/injected here -->
  <div><!-- this div is the injected parent -->
    <my-list-item virtual-repeat.for="item of items">
      ...
    </my-list-item>
  </div>
<my-list>
```

**Comment by lgabeskiria** [comment]
any progress on this?

**Comment by Thanood** [comment]
Somehow, this totally escapes me..
i just think I don't know enough about `#document-fragment`, viewSlots or whatever would help here. :smile: 

It works when wrapping the item to be repeated inside another div. Doing the solution (or hack) above manually like so:

``` html
<my-list>
  <div>
    <my-list-item virtual-repeat.for="item of items">
      ...
    </my-list-item>
  </div>
<my-list>
```

I couldn't manage doing this in another view strategy, yet. The element "disappears".. Guess, I just didn't find the right spot.
As said, I'm still learning this document-fragment stuff. Although it'd be a simple solution. :smile:

**Comment by Thanood** [comment]
@martingust please see my fork here: https://github.com/Thanood/ui-virtualization

This doesn't even work and feels ugly. :smile: 

I've basically introduced [a new view strategy](https://github.com/Thanood/ui-virtualization/blob/master/src/view-strategy.js#L88) if the parent of the `virtual-repeat` attribute's element is a `#document-fragment`. In this case, the list is surrounded with another `<div>`.

Here is the ugly part:
https://github.com/Thanood/ui-virtualization/blob/master/src/virtual-repeat.js#L86-L93

I could think of some fixes of this situation. Like introduce a general `createWrapper` function inside each strategy which will just return in everything but `DocumentFragmentStrategy`. But that's not much better. :smile: 

I'd love to see a hint here on how to do this correctly.

**Comment by martingust** [comment]
@Thanood Thanks for looking into this. I don't think adding a surrounding `div` would be necessary though. The problem is [here](https://github.com/aurelia/ui-virtualization/blob/master/src/virtual-repeat.js#L349) where the view's `firstChild` (the <view> - comment) does not have a parent element since that would be a custom element which aurelia manually inserts to the DOM. I think a solution is to look at the viewSlot's contentSelectors, find the custom element and calculate its height.

Inserting this

``` javascript
if (!firstViewElement) {
      firstViewElement = this.viewSlot.contentSelectors[0].groups[0][0];
    }
```

[here](https://github.com/aurelia/ui-virtualization/blob/master/src/virtual-repeat.js#L350) temporary fixes the problem in this particular scenario. But a more solid way to find the custom element among the contentSelectors would probably be needed.

**Comment by EisenbergEffect** [comment]
@martingust Bear in mind this has all changed with the new shadow dom. There is a release available on github. We need to try to start aligning with that for the RC. It's not available through npm so it doesn't break others, only on Github.

**Comment by Thanood** [comment]
@EisenbergEffect Your Gitter post popped up literally seconds after I posted here and I just thought you're reading my mind or something. :smile: 

@martingust Great hint, thanks!
I guess I'll try again with `<slot>` and see what happens when I use it.
If that doesn't change too much I will have a look at contentSelectors.

**Comment by martingust** [comment]
@EisenbergEffect Oh, I see that :) Will need to look into how virtual-repeat can align.

**Comment by martingust** [comment]
@Thanood @EisenbergEffect I've synced `virtual-repeat` to work with the new shadow DOM stuff. It seems to work good and also resolves this issue with custom elements.

**Comment by Thanood** [comment]
Thanks a lot! :+1: 
I will do a test run with my list later.

**Comment by JSeligsohn** [comment]
It seems I'm still running into a similar problem as @Thanood.
I have a custom element that uses `virtual-repeat` and receiving the error "Could not calculate item height" - the "parent element" it's looking at is the empty `div` that is inserted into the container to keep track of the view.

Here's the custom element:

``` html
 <product-cards products.bind="products" totalcount.one-time="totalCount"></product-cards>
```

Here is the input for the custom element:

``` html
<template>
<div class="products-scroll-container" style="overflow:scroll">
  <div virtual-repeat.for="i of numRows" class="products-row">
    <div repeat.for="product of formattedProducts | sliceIt: [i,3]" class="product-container">...</div>
  </div>
</div>
</template>
```

Here is the output:

``` html
<div class="products-scroll-container" style="overflow:scroll">
    <div></div>
    <div class="products-row">
       <div class="product-container">....</div>
       <div class="product-container">....</div>
       <div class="product-container">....</div>
    </div>
    <div></div>
</div>
```

I've tried placing the class "products-scroll-container" on the custom element itself, but receive the same error.

**Comment by Thanood** [comment]
I'm getting another error with aurelia-templating 1.0.0-beta.3.0.2.

```
aurelia-templating.js:2767 Uncaught (in promise) TypeError: Cannot read property 'attrToRemove' of undefined
    at ViewCompiler._compileElement (http://127.0.0.1:8080/jspm_packages/npm/aurelia-templating@1.0.0-beta.3.0.2/aurelia-templating.js:2767:27)
    at ViewCompiler._compileNode (http://127.0.0.1:8080/jspm_packages/npm/aurelia-templating@1.0.0-beta.3.0.2/aurelia-templating.js:2490:23)
    at ViewCompiler._compileElement (http://127.0.0.1:8080/jspm_packages/npm/aurelia-templating@1.0.0-beta.3.0.2/aurelia-templating.js:2782:31)
    at ViewCompiler._compileNode (http://127.0.0.1:8080/jspm_packages/npm/aurelia-templating@1.0.0-beta.3.0.2/aurelia-templating.js:2490:23)
    at ViewCompiler._compileNode (http://127.0.0.1:8080/jspm_packages/npm/aurelia-templating@1.0.0-beta.3.0.2/aurelia-templating.js:2512:33)
    at ViewCompiler.compile (http://127.0.0.1:8080/jspm_packages/npm/aurelia-templating@1.0.0-beta.3.0.2/aurelia-templating.js:2459:12)
    at eval (http://127.0.0.1:8080/jspm_packages/npm/aurelia-templating@1.0.0-beta.3.0.2/aurelia-templating.js:3109:50)
```

The variable `expression` here is `undefined`:
https://github.com/aurelia/templating/blob/master/src/view-compiler.js#L439

That is, the `expressions` array has one element which is `undefined`.

Repro: https://github.com/Thanood/ui-virt-expression-issue

**Comment by martingust** [comment]
@JSeligsohn @Thanood I'll investigate this.

**Comment by Daniel-Glucksman** [comment]
@martingust Is their any update on this?

**Comment by martingust** [comment]
@devteamexpress Sorry for the delay, other stuff have been prioritized. My plan is to start knock out issues (this included) on this repo within a week or so.

**Comment by martingust** [comment]
@JSeligsohn The `virtual-repeat` attribute needs to be placed on an element that has a height the browser can calculate, either a set height with CSS or some content in the html. In your example the content in your (inner) `repeat.for` has not been rendered yet and the element (div.product-row) might not have a set height with CSS?

**Comment by martingust** [comment]
@Thanood Unfortunately got error when running gulp in your repro project, but I did copy all your files into UI-Virtualization's sample project and it all ran without errors.

**Comment by Thanood** [comment]
Thanks, @martingust. I will try this again with templating RC1 and report back.
Will also have a look at this gulp error. :)

**Comment by Thanood** [comment]
@martingust No errors with `npm:aurelia-templating@^1.0.0-rc.1.0.0`! :smile: 

Scrolling is "strange", though. Here's an animated gif showing what I mean:

http://i.imgur.com/Y3cBGK1.gifv
(it takes a few seconds before scrolling starts)

Please note that this list has 1000 rows. The template used is pretty simple:

``` html
<template>
    <div>
        <my-list>
            <my-list-item virtual-repeat.for="item of data">
                ${item.title}
            </my-list-item>
        </my-list>
    </div>
</template>
```

Also I didn't experience any gulp errors. :smile: 

But maybe this is another issue. The issue here "problems with custom elements" seems to be solved. :+1:

**Comment by martingust** [comment]
@Thanood Good that worked. I see the issue in the git and it looks very much like the same issue as this https://github.com/aurelia/ui-virtualization/issues/68. I'll close this issue, and will investigete the other issue.

**Comment by martingust** [comment]
@Thanood I located and resolved #68 but indeed it was not the same issue. There is a problem to insert the buffer elements in conjunction with content projection - right now it tries to insert the buffer elements before `au-content` element which is not in the DOM at the time and the buffer end up not being added to the DOM. Will need to look more into this.
